gneiss/util.py

```diff
@@ -317,3 +317,74 @@ def random_tree(n):
     t = TreeNode.from_linkage_matrix(lm, ids)
     t = rename_internal_nodes(t)
     return t
+
+
+def block_diagonal(ncols, nrows, nblocks):
+    """ Generate block diagonal with uniformly distributed values within blocks.
+
+    Parameters
+    ----------
+    ncol : int
+        Number of columns
+    nrows : int
+        Number of rows
+    nblocks : int
+        Number of blocks
+    """
+    mat = np.zeros((nrows, ncols))
+    block_cols = ncols // nblocks
+    block_rows = nrows // nblocks
+    for b in range(nblocks-1):
+        B = np.random.uniform(size=(block_rows, block_cols))
+        lower_row = block_rows * b
+        upper_row = min(block_rows*(b+1), nrows)
+        lower_col = block_cols * b
+        upper_col = min(block_cols*(b+1), ncols)
+
+        mat[lower_row:upper_row, lower_col:upper_col] = B
+
+    # Make last block fill in the remainder
+    B = np.random.uniform(size=(nrows-upper_row, ncols-upper_col))
+    mat[upper_row:, upper_col:] = B
+    return mat
+
+
+def _shift(l, n):
```

Comment:
can an english description of this algorithm be provided? is it just rolling the vector? if so, can `np.roll` be used?